Add one-pager about the switching to a new underlying test framework (#11)
crossplane/uptest: design/one-pager-considerations-for-changing-test-framework-of-uptest.md

# Considerations for Changing Test Framework of Uptest

* Owner: Sergen Yalcin (@sergenyalcin)
* Reviewers: Uptest Maintainers
* Status: Draft

## Background

Uptest is a tool for testing and validating Crossplane providers. This tool is
utilized to test the behavior of Crossplane controllers, resource management
operations, and other Crossplane components. Uptest can simulate the creation,
update, import, deletion, and other operations of Crossplane resources.

When `Uptest` was first written, it used [kuttl](https://github.com/kudobuilder/kuttl)
as the underlying test framework in order to have good declarative testing
capabilities. Over time, it was realized that there are better and more

> compatible alternatives and `kuttl` is a tool that is not currently maintained,
> so it was decided to change the underlying framework.

Reviewer: ```suggestion
compatible alternatives and a perception that `kuttl` isn't being actively maintained,
so it was decided to evaluate an alternative underlying framework.
```

[kuttl](https://github.com/kudobuilder/kuttl) provides a declarative approach to
test Kubernetes Operators. `kuttl` is designed for testing operators, however it
can declaratively test any kubernetes objects.

> While starting to the `Uptest` effort, we considered a few different
> alternatives and `kuttl`'s capabilities were appropriate for our assertion
> aims although its missing points. Today, we consider changing the underlying
> test framework tool because of the `kuttl` stopped to be actively maintained.

Reviewer: ```suggestion
When starting the `Uptest` effort, we considered a few different
alternatives and `kuttl`'s capabilities were appropriate for our assertion
aims even though it missed some points. Today, we consider changing the underlying
test framework tool because of the perception of `kuttl` not being actively maintained
and other frameworks offering superior capabilities.
```

Author: Done!

> ## Goals
> 
> - Determining the requirements of Uptest well and revealing its needs.
> - Having a more inclusive and capable underlying test framework.
> - Use of a more up-to-date and maintained tool in the background.

Reviewer: ```suggestion
## Goal

Decide on a more comprehensive underlying test framework to meet the current and [future requirements](https://github.com/crossplane/uptest/pull/10/files) of Uptest.
```

Author: Done!

> ## Proposal - Switching to `chainsaw` tool

Reviewer: ```suggestion
## Proposal - Switching to `chainsaw`
```

Author: Done!

[chainsaw](https://github.com/kyverno/chainsaw) provides a declarative approach
to test Kubernetes operators and controllers. While Chainsaw is designed for
testing operators and controllers, it can declaratively test any Kubernetes
objects. Chainsaw is an open-source tool that was initially developed for
defining and running Kyverno end-to-end tests. The tool has Apache-2.0 license.

> In addition to the functionality provided by `kuttl`, the prominent features of
> the `chainsaw` tool are that it is maintained up to date and has an
> infrastructure for migration from `kuttl`.

Reviewer: ```suggestion
In addition to providing similar functionality provided by `kuttl`,  it also offers better logs, config maps assertions, [assertions trees](https://kyverno.io/blog/2023/12/13/kyverno-chainsaw-exploring-the-power-of-assertion-trees/) and many more things. The fact that it is well-maintained, and has the capability for migration from `kuttl` makes it an attractive option.
```

Author: Done!

> There are very similar concepts in `chainsaw` with `kuttl`. By this way, we do
> not have to make important changes in the templates.

Reviewer: ```suggestion
`chainsaw` shares similar concepts with `kuttl`. In this way, we do
not have to make major changes to the templates.
```

Author: Done!

    # ...
```

```yaml
apiVersion: chainsaw.kyverno.io/v1alpha1
kind: Test
metadata:
  name: example
spec:
  steps:
  - try:
    # ...
    - command:
        entrypoint: echo
        args:
        - hello chainsaw
    # ...
```

Also provides logical assertion statements:

```yaml
apiVersion: chainsaw.kyverno.io/v1alpha1
kind: Test
metadata:
  name: example
spec:
  steps:
  - try:
    # ...
    - assert:
        resource:
          apiVersion: v1
          kind: Deployment
          metadata:
            name: foo
          spec:
            (replicas > 3): true
    # ...
```

Resource Template support is another important requirement for Uptest:

```yaml
apiVersion: chainsaw.kyverno.io/v1alpha1
kind: Test
metadata:
  name: template
spec:
  template: true
  steps:
  - assert:
      resource:
        # apiVersion, kind, name, namespace and labels are considered for templating
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: ($namespace)
        # other fields are not (they are part of the assertion tree)
        data:
          foo: ($namespace)
```

Related Issue: https://github.com/upbound/official-providers-ci/issues/179

> In short, the `chainsaw` is more capable and maintained tool than `kuttl`. So,
> switching to this new tool may better for handling the Uptest requirements.

Reviewer: ```suggestion
In short, `chainsaw` is a more capable and well-maintained framework than `kuttl` and switching to it will better suit Uptest's future requirements.
```

## Alternative Considered

### Using the `crossplane-e2e-framework`

[crossplane-e2e-framework](https://github.com/crossplane/crossplane/tree/master/test/e2e)
is a [k8s-e2e-framework](https://pkg.go.dev/sigs.k8s.io/e2e-framework)-based
test framework that provides a baseline for simulating the real-world use to
exercise all of `Crossplane`'s functionality.

`e2e-framework` is a tool that allows tests to be written in Go. Additionally,
one of its advantages is that it works with familiar conventions in the
environment we use. On the other hand, these types of utilities can be used when
writing tests, thanks to their functions specific to the crossplane ecosystem.
However, this will mean changing the entire `Uptest` code-base currently used.
In this context, it should be taken into consideration that such a change would
be quite large.

Based on the discussion here, an analysis can be made to clarify the use cases
of `Uptest` and the `e2e-framework` and which one the user would prefer

> according to their needs through documentation.

Reviewer: I'm not following what is intended with this paragraph?

Author: I forgot to refer to the discussion. Added link. Also extended the expression.
